frontend/src/api/api.ts

```diff
@@ -0,0 +1,203 @@
+            ? hashOrPublicKey
+            : undefined,
+          mainPurse: account.mainPurse,
+          rawAccount: JSON.stringify(account),
+        };
+      },
+    },
+    validator: {
+      async getValidators() {
+        type Response = AxiosResponse<ApiData.Validators>;
+
+        const response = await middlewareApi.get<Response>('/validators');
+
+        if (response.status !== 200) throw new Error(response.statusText);
+
+        const {
+          data: { validators },
+        } = response;
+
+        return validators;
+      },
+    },
+    /**
+     * Retrieve an aggregate of the various states a deploy goes through, given its deploy hash. The node does not emit this event, but the Sidecar computes it and returns it for the given deploy. This endpoint behaves differently than other endpoints, which return the raw event received from the node.
+     * @param hash deploy hash to get deploy
+     */
+    // TODO: should we perform the business logic in the redux action?
+    deploy: {
+      async getDeploy(hash: string) {
+        type Response = AxiosResponse<ApiData.Deploy>;
+
+        const response = await middlewareApi.get<Response>(`/deploy/${hash}`);
+
+        if (response.status !== 200) throw new Error(response.statusText);
+
+        const {
+          data: { execution_results: executionResults, ...deploy },
+        } = response;
+
+        // @ts-ignore
+        const paymentMap = new Map(deploy.payment.ModuleBytes?.args);
+
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call
+        const paymentAmount = CLValueParsers.fromJSON(paymentMap.get('amount'))
+          .unwrap()
+          .value()
+          .toString() as string;
+
+        const { timestamp, account: publicKey } = deploy.header;
+
+        const { block_hash: blockHash, result: executionResult } =
+          executionResults[0];
+
+        const status = executionResult.Success
+          ? DeployStatus.Success
+          : DeployStatus.Failed;
+
+        const deploySession = deploy.session as JsonDeploySession;
+
+        const { action, deployType, amount } = determineDeploySessionData(
+          deploySession,
+          status,
+        );
+
+        const cost = executionResult.Success
+          ? executionResult.Success.cost
+          : executionResult.Failure?.cost ?? 0;
+
+        const dateTime = new Date(timestamp);
+
+        const timeSince = formatTimeAgo(dateTime);
+        const readableTimestamp = formatDate(dateTime);
+
+        return {
+          timestamp,
+          timeSince,
+          readableTimestamp,
+          deployHash: deploy.hash,
+          blockHash,
+          publicKey,
+          action,
+          deployType,
+          amount,
+          paymentAmount,
+          cost: cost.toString(),
+          status,
+          rawDeploy: JSON.stringify({
+            deploy,
+            execution_results: executionResults,
+          }),
+        };
```

Comment:
@hoffmannjan Do you think this logic could exist within the middleware or somewhere else?  I'm just thinking it'd be a challenge to test this code

Reply:
Hahaha, consistency camp 😅